File: cpp/say/say.cpp
  #include "say.h"

#include <string>

std::string say::in_english(unsigned long long input) {
  // Check input is larger than 0.
  unsigned long long oneTrillion = 1000ULL*1000ULL*1000ULL*1000ULL;
  if (input < 0 || input >= oneTrillion) throw std::domain_error("Bad input");

  // Return "zero" if 0;
  if (input == 0) return "zero";

  // Arrays with words.
  std::string upToTen[] = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
  std::string tenToTwenty[] = {"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "eighteen", "nineteen"};
  std::string tensMultiple[] = {"", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};

  // Ouptut string
  std::string words = "";

  // Billions.
  if (input / 1000000000 > 0) {
    words += in_english(input / 1000000000) + " billion";
    input %= 1000000000;
    // Add trailing space if input still > 0.
    if (input > 0) words += " ";
  }

  // Millions.
  if (input / 1000000 > 0) {
    words += in_english(input / 1000000) + " million";
    input %= 1000000;
    // Add trailing space if input still > 0.
    if (input > 0) words += " ";
  }

  // Thousands.
  if (input / 1000 > 0) {
    words += in_english(input / 1000) + " thousand";
    input %= 1000;
    // Add trailing space if input still > 0.
    if (input > 0) words += " ";
  }

  // Hundreds
  if (input / 100 > 0) {
    words += in_english(input / 100) + " hundred";
    input %=100;
    // Add trailing space if input still > 0.
    if (input > 0) words += " ";
  }

  // Between 10 and 20.
  if (input > 10 && input < 20) {
    words += tenToTwenty[input%10];
  } else if (input >= 20) {
    words += tensMultiple[input/10];
    input %= 10;
    // Add '-' for final word if input still > 0.
    if (input > 0) words += "-";
  }

  if (input < 10 && input > 0) {
    words += upToTen[input];
  }

  return words;
}

File: clojure/say/src/say.clj
(ns say)

(def belowTen {1 "one", 2 "two", 3 "three", 4 "four", 5 "five", 6 "seven",
               8 "eight", 9 "nine"})
(def belowTwenty {10 "ten", 11 "eleven", 12 "twelve", 13 "thirteen",
                  14 "fourteen", 15 "fifteen", 16 "sixteen", 17 "seventeen",
                  18 "eighteen", 19 "nineteen"})
(def tens {20 "twenty", 30 "thirty", 40 "forty", 50 "fifty", 60 "sixty",
           70 "seventy", 80 "eighty", 90 "ninety"})

(defn deconstruct-number
  [num]
  (->> num
       (str)
       (map #(Character/digit % 10))
       (split-with (count >= 3))))


(defn construct-number-string
  ([num s] (if (zero? num)
             "zero"
             (->> (cond
                    (< num 10) (belowTen num)
                    (< num 20) (belowTwenty num)
                    (< num 100) (tens num))
                  (str s)))))


(defn number
  [num]
  (construct-number-string num ""))
